plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationFieldMapper.java

```
@@ -0,0 +1,623 @@
+
+        public CollationFieldType clone() {
+            return new CollationFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder() {
+            failIfNoDocValues();
+            return new DocValuesIndexFieldData.Builder();
+        }
+
+        public static DocValueFormat COLLATE_FORMAT = new DocValueFormat() {
+            @Override
+            public String getWriteableName() {
+                return "collate";
+            }
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
+            }
+
+            @Override
+            public String format(long value) {
+                return Long.toString(value);
+            }
+
+            @Override
+            public String format(double value) {
+                return Double.toString(value);
+            }
+
+            @Override
+            public String format(BytesRef value) {
+                int encodedLength = IndexableBinaryStringTools.getEncodedLength(value.bytes, value.offset, value.length);
+                char[] encoded = new char[encodedLength];
+                IndexableBinaryStringTools.encode(value.bytes, value.offset, value.length, encoded, 0, encodedLength);
+                return new String(encoded, 0, encodedLength);
+            }
+
+            @Override
+            public long parseLong(String value, boolean roundUp, LongSupplier now) {
+                double d = Double.parseDouble(value);
+                if (roundUp) {
+                    d = Math.ceil(d);
+                } else {
+                    d = Math.floor(d);
+                }
+                return Math.round(d);
+            }
+
+            @Override
+            public double parseDouble(String value, boolean roundUp, LongSupplier now) {
+                return Double.parseDouble(value);
+            }
```

parseLong and parseDouble should just throw an exception